presto-main/src/main/java/com/facebook/presto/util/KerberosPrincipal.java

```diff
@@ -0,0 +1,201 @@
+
+            usernameAndHostname = value.substring(0, realmSeparatorIndex);
+            realm = Optional.of(value.substring(realmSeparatorIndex + 1, value.length()));
+        }
+        else {
+            usernameAndHostname = value;
+            realm = Optional.empty();
+        }
+
+        String username;
+        Optional<String> hostname;
+        if (hostnameSeparatorIndex >= 0) {
+            // username cannot be empty
+            if (hostnameSeparatorIndex == 0) {
+                throw invalidKerberosPrincipal(value);
+            }
+
+            // hostname cannot be empty
+            if (hostnameSeparatorIndex == usernameAndHostname.length() - 1) {
+                throw invalidKerberosPrincipal(value);
+            }
+
+            username = usernameAndHostname.substring(0, hostnameSeparatorIndex);
+            hostname = Optional.of(usernameAndHostname.substring(hostnameSeparatorIndex + 1, usernameAndHostname.length()));
+        }
+        else {
+            username = usernameAndHostname;
+            hostname = Optional.empty();
+        }
+
+        return new KerberosPrincipal(username, hostname, realm);
+    }
+
+    private static IllegalArgumentException invalidKerberosPrincipal(String value)
+    {
+        return new IllegalArgumentException("Invalid Kerberos principal: " + value);
+    }
+
+    public KerberosPrincipal(String userName, Optional<String> hostName, Optional<String> realm)
+    {
+        this.username = requireNonNull(userName, "username is null");
+        this.hostname = requireNonNull(hostName, "hostname is null");
+        this.realm = requireNonNull(realm, "realm is null");
+    }
+
+    public String getUserName()
+    {
+        return username;
+    }
+
+    public Optional<String> getHostName()
+    {
+        return hostname;
+    }
+
+    public Optional<String> getRealm()
+    {
+        return realm;
+    }
+
+    public KerberosPrincipal substituteHostnamePlaceholder()
+    {
+        return substituteHostnamePlaceholder(getLocalCanonicalHostName());
+    }
+
+    public KerberosPrincipal substituteHostnamePlaceholder(String hostname)
+    {
+        if (this.hostname.isPresent() && this.hostname.get().equals(HOSTNAME_PLACEHOLDER)) {
+            return new KerberosPrincipal(username, Optional.of(hostname), realm);
+        }
+        return this;
+    }
+
+    private static String getLocalCanonicalHostName()
+    {
+        try {
+            return InetAddress.getLocalHost().getCanonicalHostName().toLowerCase(Locale.US);
+        }
+        catch (UnknownHostException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder builder = new StringBuilder();
+        builder.append(username);
+        hostname.ifPresent(hostname -> builder.append(PARTS_SEPARATOR).append(hostname));
+        realm.ifPresent(realm -> builder.append(REALM_SEPARATOR).append(realm));
+        return builder.toString();
```

Comment:
If `hostname` or `realm` uses `PARTS_SEPARATOR` or `REAL_SEPARATOR` in its value, don't you need to use `ESCAPE_CHARACTER` there?